### PR TITLE
Support global properties for azure active directory

### LINKED ISSUE #25646: [FEATURE REQ] In aad-starter support properties like spring.cloud.azure.tenant-id
If spring.cloud.azure.active-directory.tenant-id is not configured, use spring.cloud.azure.tenant-id instead.
Same to other properties like client-id, client-secrets.

### PR DESCRIPTION
**Pull request by chenrujun**
Resolve #25646

### CODE REVIEW

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/properties/AADPropertiesConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+@Configuration
+@EnableConfigurationProperties
+public class AADPropertiesConfiguration implements InitializingBean {
+
+    @Bean
+    @ConditionalOnMissingBean
+    @ConfigurationProperties(prefix = AzureGlobalProperties.PREFIX)
+    public AzureGlobalProperties azureGlobalProperties() {
+        return new AzureGlobalProperties();
+    }
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/properties/AADPropertiesConfiguration.java:26]: There's already one AzureGlobalProperties bean, no need to configure it another time

**Review by saragluna — CHANGES_REQUESTED** [review]

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/properties/AADPropertiesConfiguration.java (hunk 1 above)
Comment (chenrujun) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/properties/AADPropertiesConfiguration.java]: Updated.

**Comment by chenrujun** [comment]
Hi, @saragluna , @stliu  , @backwind1233 , please help to review this PR when you have time.

Here is related sample PR: https://github.com/Azure-Samples/azure-spring-boot-samples/pull/105

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/AADAutoConfiguration.java [hunk 2]
```diff
@@ -20,15 +21,16 @@
  */
 @Configuration
 @ConditionalOnProperty(value = "spring.cloud.azure.active-directory.enabled", havingValue = "true")
-@EnableConfigurationProperties({
-    AADAuthenticationProperties.class,
-    AADResourceServerProperties.class
-})
 @Import({
+    AzureGlobalPropertiesAutoConfiguration.class,
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/AADAutoConfiguration.java:25]: Why import this configuration here?

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/filter/AADAuthenticationFilterAutoConfiguration.java [hunk 3]
```diff
@@ -35,7 +37,8 @@
 @ConditionalOnWebApplication
 @ConditionalOnExpression("${spring.cloud.azure.active-directory.enabled:false}")
 @ConditionalOnMissingClass({ "org.springframework.security.oauth2.server.resource.BearerTokenAuthenticationToken" })
-@ConditionalOnProperty(prefix = AADAuthenticationFilterAutoConfiguration.PROPERTY_PREFIX, value = { "client-id" })
+@ConditionalOnProperty(prefix = AADAuthenticationFilterAutoConfiguration.PROPERTY_PREFIX, value = { "credential.client-id" })
+@Import(AzureGlobalPropertiesAutoConfiguration.class)
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/filter/AADAuthenticationFilterAutoConfiguration.java:41]: No need to import this auto-configuration here.

**Review by saragluna — CHANGES_REQUESTED** [review]

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/AADAutoConfiguration.java (hunk 2 above)
Comment (chenrujun) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/AADAutoConfiguration.java:25]: There will be `NoSuchBeanDefinitionException` error if we delete this.

image


image

Comment (saragluna) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/AADAutoConfiguration.java:25]: use constructor injection in the aad autoconfiguration class, and provide a bean in your test.

**Comment by chenrujun** [comment]
@saragluna , updated.

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/properties/AADAuthenticationProperties.java [hunk 4]
```diff
@@ -695,40 +617,45 @@ public boolean isAllowedGroup(String group) {
...
     }
 
-    @Override
-    public void afterPropertiesSet() {
-
-        if (!StringUtils.hasText(baseUri)) {
-            baseUri = "https://login.microsoftonline.com/";
-        } else {
-            baseUri = addSlash(baseUri);
+    private void setDefaultValueFromAzureGlobalProperties(AzureGlobalProperties global) {
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/properties/AADAuthenticationProperties.java:620]: There's another way to do this, which will provide the default value before binding. So there's no need to autowire the globalAzureProperties in this configuration properties. 
```java
@ConfigurationProperties(prefix = AADAuthenticationProperties.PREFIX)
@Bean
public AADAuthenticationProperties aadAuthenticationProperties() {
    return loadProperties(this.azureGlobalProperties, new AADAuthenticationProperties());
}
```

**Review by saragluna — CHANGES_REQUESTED** [review]

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/properties/AADAuthenticationProperties.java (hunk 4 above)
Comment (chenrujun) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/properties/AADAuthenticationProperties.java]: @saragluna , updated.

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/properties/AADAuthenticationProperties.java [hunk 5]
```diff
@@ -120,20 +116,30 @@
     private boolean allowTelemetry = true;
 
     /**
-     * If <code>true</code> activates the stateless auth filter {@link AADAppRoleStatelessAuthenticationFilter}. The
-     * default is <code>false</code> which activates {@link AADAuthenticationFilter}.
+     * If true activates the stateless auth filter {@link AADAppRoleStatelessAuthenticationFilter}. The
+     * default is false which activates {@link AADAuthenticationFilter}.
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/properties/AADAuthenticationProperties.java:120]: should we also remove the {@link } here?

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/resources/META-INF/additional-spring-configuration-metadata.json [hunk 6]
```diff
@@ -230,7 +230,7 @@
       "deprecation": {
         "level": "error",
         "reason": "Use uniform property format(spring.cloud.azure.*) in all Spring Cloud Azure libraries.",
-        "replacement": "spring.cloud.azure.active-directory.base-uri"
+        "replacement": "spring.cloud.azure.active-directory.profile.environment.active-directory-endpoint"
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/resources/META-INF/additional-spring-configuration-metadata.json:233]: Are we also going to update the migration guide?

**Review by saragluna — CHANGES_REQUESTED** [review]

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/resources/META-INF/additional-spring-configuration-metadata.json (hunk 6 above)
Comment (chenrujun) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/resources/META-INF/additional-spring-configuration-metadata.json:233]: Already created a PR: https://github.com/microsoft/spring-cloud-azure/pull/786

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/properties/AADAuthenticationProperties.java (hunk 5 above)
Comment (chenrujun) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/properties/AADAuthenticationProperties.java:120]: @Why?

Comment (chenrujun) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/properties/AADAuthenticationProperties.java]: Updated.

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/properties/AADAuthenticationProperties.java [hunk 7]
```diff
@@ -120,20 +116,30 @@
     private boolean allowTelemetry = true;
 
     /**
-     * If <code>true</code> activates the stateless auth filter {@link AADAppRoleStatelessAuthenticationFilter}. The
-     * default is <code>false</code> which activates {@link AADAuthenticationFilter}.
+     * If true activates the stateless auth filter AADAppRoleStatelessAuthenticationFilter. The
+     * default is false which activates  AADAuthenticationFilter.
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/properties/AADAuthenticationProperties.java:120]: nit: two spaces before AADAuthenticationFilter.

**Review by saragluna — APPROVED** [review]
LGTM

**Pull request merged**